### ISSUE TITLE
Segmentation fault when `fl::*` objects embedded in C++ class

### ISSUE THREAD

**Issue by rayvburn** [fuzzylite/fuzzylite#93]
Hello, I'm struggling with a problem of `Segmentation fault (core dumped)` when I embed `fl::Engine` etc. inside C++ class. No problems with objects dynamically allocated outside class (even if I don't explicitly `delete` them at the end).

I've prepared a small example that allows to reproduce this problem (`gtest` required).
Is that my FIS configuration issue or some kind of library problem? I haven't tried to define the same FIS in `FuzzyLite Language`, because this is not a solution in my application - I need to dynamically define `fl::Trapezoid` vertices.

`test_debug_fis.cpp`:
```c++
#include <gtest/gtest.h>
#include <fl/Headers.h>

void configure(
	fl::Engine* engine_ptr,
	fl::InputVariable* in_l_ptr,
	fl::InputVariable* in_d_ptr,
	fl::OutputVariable* out_a_ptr,
	fl::RuleBlock* rule_block_ptr
) {
	/* Initialize engine */
	engine_ptr->setName("test");
	engine_ptr->setDescription("");

	/* Initialize first input variable */
	in_l_ptr->setName("l");
	in_l_ptr->setDescription("");
	in_l_ptr->setEnabled(true);
	in_l_ptr->setRange(-3.1415, +3.1415);
	in_l_ptr->setLockValueInRange(false);
	in_l_ptr->addTerm(new fl::Triangle("lba",  -3.1415, -3.1415, -2.7925));
	engine_ptr->addInputVariable(in_l_ptr);

	/* Initialize second input variable */
	in_d_ptr->setName("d");
	in_d_ptr->setDescription("");
	in_d_ptr->setEnabled(true);
	in_d_ptr->setRange(-3.1415, +3.1415);
	in_d_ptr->setLockValueInRange(false);
	in_d_ptr->addTerm(new fl::Trapezoid("do",  -3.1415, -2.6180, -2.0944, -1.5708));
	engine_ptr->addInputVariable(in_d_ptr);

	/* Initialize output variable */
	out_a_ptr->setName("a");
	out_a_ptr->setDescription("");
	out_a_ptr->setEnabled(true);
	out_a_ptr->setRange(-3.1415, +3.1415);
	out_a_ptr->setLockValueInRange(false);
	out_a_ptr->setAggregation(new fl::Maximum);
	out_a_ptr->setDefuzzifier(new fl::Centroid());
	out_a_ptr->setDefaultValue(fl::nan);
	out_a_ptr->setLockPreviousValue(false);

	out_a_ptr->addTerm(new fl::Trapezoid("aa", -0.7070, -0.3501, +0.3501, +0.7070));
	engine_ptr->addOutputVariable(out_a_ptr);

	rule_block_ptr->setName("mamdani");
	rule_block_ptr->setDescription("");
	rule_block_ptr->setEnabled(true);
	rule_block_ptr->setConjunction(new fl::Minimum);
	rule_block_ptr->setDisjunction(new fl::AlgebraicSum);
	rule_block_ptr->setImplication(new fl::Minimum);
	rule_block_ptr->setActivation(new fl::General);

	rule_block_ptr->addRule(fl::Rule::parse("if l is lba and d is do then a is aa", engine_ptr));
}

// class
class FlEmbedded {
public:
	FlEmbedded();
	virtual ~FlEmbedded();
private:
	fl::Engine* engine_ptr_;
	fl::InputVariable* in_l_ptr_;
	fl::InputVariable* in_d_ptr_;
	fl::OutputVariable* out_a_ptr_;
	fl::RuleBlock* rule_block_ptr_;
};

FlEmbedded::FlEmbedded():
	engine_ptr_(new fl::Engine()),
	in_l_ptr_(new fl::InputVariable()),
	in_d_ptr_(new fl::InputVariable()),
	out_a_ptr_(new fl::OutputVariable()),
	rule_block_ptr_(new fl::RuleBlock())
{
	configure(engine_ptr_, in_l_ptr_, in_d_ptr_, out_a_ptr_, rule_block_ptr_);
	try {
		engine_ptr_->addRuleBlock(rule_block_ptr_);
	} catch (fl::Exception& what) {
		std::cout << what.what() << std::endl;
	}
}

FlEmbedded::~FlEmbedded() {
	delete engine_ptr_;
	delete in_l_ptr_;
	delete in_d_ptr_;
	delete out_a_ptr_;
	delete rule_block_ptr_;
}

TEST(FuzzyInferenceSystem, outsideClass) {
	fl::Engine* engine_ptr_ = new fl::Engine();
	fl::InputVariable* in_l_ptr_ = new fl::InputVariable();
	fl::InputVariable* in_d_ptr_ = new fl::InputVariable();
	fl::OutputVariable* out_a_ptr_ = new fl::OutputVariable();
	fl::RuleBlock* rule_block_ptr_ = new fl::RuleBlock();

	configure(engine_ptr_, in_l_ptr_, in_d_ptr_, out_a_ptr_, rule_block_ptr_);
	try {
		engine_ptr_->addRuleBlock(rule_block_ptr_);
	} catch (fl::Exception& what) {
		std::cout << what.what() << std::endl;
	}
}

TEST(FuzzyInferenceSystem, embeddedInClass) {
	FlEmbedded object;
}

int main(int argc, char** argv) {
	testing::InitGoogleTest(&argc, argv);
	return RUN_ALL_TESTS();
}
```

Test output (results not created):
```console
Segmentation fault (core dumped)                                                                                                                                                
[{some_pkg}:make] -- run_tests.py: verify result "{some_path}/gtest-test_debug_fis.xml"
[{some_pkg}:make] Cannot find results, writing failure results to '{some_path}/MISSING-gtest-test_debug_fis.xml'
```

I'm running that on Ubuntu 16.04 machine. Tests are run as a part of Robot Operating System package, if that matters (ROS Kinetic, `catkin` used as a workspace tool).

I will appreciate your support.

**Comment by jcrada** [comment]
Hi,

Thanks for your detailed post.

The problem you have is that the engine components (input and output variables and rule blocks) are managed by the engine they are in, so they get deleted with the engine.

In your case, you delete the engine and then its components in the embedded destructor, but the components are already deleted when you first delete the engine.

Check the engine source code and let me know if you have further questions.

Cheers

**Comment by rayvburn** [comment]
Thanks for a quick answer. I checked `Engine` destructor and noticed that it does [delete](https://github.com/fuzzylite/fuzzylite/blob/7aee562d6ca17f3cf42588ffb5116e03017c3c50/fuzzylite/src/Engine.cpp#L96) inputs, outputs and rule blocks indeed. I changed `FlEmbedded` destructor to simply delete only `engine_ptr_` and the problem has gone.

```c++
FlEmbedded::~FlEmbedded() {
	delete engine_ptr_;
}
```

Thank you for your effort!
